### ISSUE TITLE
Morris method input

### ISSUE THREAD

**Issue by vijaya65** [JuliaDiffEq/DifferentialEquations.jl#575]
I have a question regarding the Morris method in the  differential equation package.
The documentation does not describe how the number of trajectories is calculated for the Morris method.
In the example given for the LotkaVolterra model with 4 parameters, an unusually large number of trajectories
is included in the input

m = gsa(f1,Morris(total _num_trajectory=1000,num_trajectory=150),[[1,5],[1,5],[1,5],[1,5]])

From what I have read about the method for 4 parameters the number of trajectories would be 5x(r=8) or 40 trajectories at most. 
Do you know how the numbers 1000 and 150 are calculated for the Lotka Volterra model and what they represent?

**Comment by ChrisRackauckas** [comment]
@Vaibhavdixit02 can comment.

In the Morris method, you take independent trajectories in order to chain together evaluations. IIRC, here we are taking 1000 trajectories and 150 evaluations along each trajectory, so 1000*150 total evaluations. The number of parameters does not factor into the trajectory length in a Morris run, but in your head you should remember that to span the space you should grow it polynomially with the number of parameters.

**Comment by vijaya65** [comment]
Hello Chris,

Every article including Matlab and R documentation say something different from your description below.
Wikipedia also has a simple and easy to follow description under elementary effects.
Or this paper
A review on global sensitivity analysis methodsBertrand Iooss12and Paul Lemaˆıtre131EDF R&D, 6 Quai Watier - 78401 Chatou, France2Institut de Math ́ematiques de Toulouse, 118 route de Narbonne - 31062 Toulouse,France3INRIA Sud-Ouest, 351 cours de la lib ́eration - 33405 Talence, France.
Or this one
Using the Morris sensitivity analysis method to assess the importance of input variables on time-reversal imaging of seismic sources

  *   Anna Franczyk<https://link.springer.com/article/10.1007/s11600-019-00356-5#auth-1>

Acta Geophysica<https://link.springer.com/journal/11600> volume 67, pages1525–1533(2019)Cite this article<https://link.springer.com/article/10.1007/s11600-019-00356-5#citeas>

  *   476 Accesses
  *   Metrics <https://link.springer.com/article/10.1007%2Fs11600-019-00356-5/metrics>



I am assuming Vaibhav set it up so lendesignmatrix=(p+1)x(num_trajectories) where p+1=the number of parameters (called variables/factors by different authors)  and num_trajectories is r in most references.

Is there any way I can confirm before using the code? In the example he did not assign a value to lendesignmatrix.  Of course if you still think I am wrong please let me know.

Thanks
Vijaya


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by Vaibhavdixit02** [comment]
Hi @vijaya65, there is no calculation of number of trajectories being done on our end, you can pass in the number of trajectories to use as per the docs.

>The total number of design matrices that are generated out of which `num_trajectory` matrices with the highest spread are used in calculation.

**Comment by vijaya65** [comment]
Thanks vaibhav, what's the length of the design matrix

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Vaibhavdixit02** [comment]
You can pass in the `len_design_mat` argument to specify the the size of the design matrix to be used. Does that make things clear for you?

**Comment by vijaya65** [comment]
No it doesn't make it clear as yet as you haven't defined the design matrix used here.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Vaibhavdixit02** [comment]
I see, so design matrix is the arrangement of the parameter values in a matrix form for better representation and computational advantage. In the design matrix each row is one step in the "trajectory" of the parameters, at each step a perturbation being added to one parameter at a time (OAT).

**Comment by vijaya65** [comment]
Yes Vaibhav, so design matrix of dimensions (num_trajectories , p+1parameters). Is that how you set it up?
Thanks

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Vaibhavdixit02** [comment]
No, it would be (len_design_matrix, p-parameters) and `num_trajectory` such design matrices are selected out of `total_num_trajectories`.

**Comment by vijaya65** [comment]
Hi
So for the Morris method len_design_matrix would be p-parameters+1? That would correspond to the start set of parameters + each additional row having one parameter changed(oat)
Thanks
Vijaya

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Vaibhavdixit02** [comment]
I meant "p" which is the number of parameters. Does that clarify things? 

I think we should continue this conversation in Slack. If you are stuck with a specific problem I would be happy to help you out on slack.

**Comment by ChrisRackauckas** [comment]
>So for the Morris method len_design_matrix would be p-parameters+1? That would correspond to the start set of parameters + each additional row having one parameter changed(oat)

No, because the choice of the next parameter bump is stochastic. Otherwise Morris would be biased and sample chords instead of the regular volume. It seems whatever source you're using is just describing a biased procedure.

**Comment by Vaibhavdixit02** [comment]
I agree with Chris, I have seen the suggestions regarding number and size of design matrix in papers that only works well in that case because of "easy" (mostly analytical) problems, when you have an actual differential equation model or something along those lines it takes a lot more number of evaluations to reach convergence. So I would suggest instead of using those heuristics to decide the arguments run the GSA a few times with different set of parameters and settle down into a combination that gives you the error scales you want.

**Comment by vijaya65** [comment]
Okay I see your point but I thought stochasticity was taken care of by running num_trajectories. Are you saying len_design_matrix can take any value or is it some multiple of p (that is number of parameters)? For example what did you use in the Lotka Volterra problem in the docs.
Was your source for your code the 2007 paper by Campolongo et al.

Thanks and hopefully this should take care of all my doubts

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by vijaya65** [comment]
Hello Vaibhav and Chris,
I tried to implement the example gsa problem in the docs:
https://docs.juliadiffeq.org/latest/analysis/global_sensitivity/#GSA-examples-1
but get the following error message:
undefVarerror:mean not defined

Is there some other statistics package that I have to install in addition to DiffEqSensitivity and DifferentialEquations to run the sensitivity analysis?
Also you had mentioned communication on another thread (slack?) how do I do that?

Thanks
Vijaya
________________________________

**Comment by Vaibhavdixit02** [comment]
Yes, for the example you would need to have `using Statistics` as well, the example already has that as the first code snippet though, it might have slipped your notice. 

`using DiffEqSensitivity, Statistics, OrdinaryDiffEq #load packages` 

@ChrisRackauckas I think we can close this issue now. Do you think we should move this to DiffEqSensitivity for posterity?

**Comment by vijaya65** [comment]
This functionality does not come standard with DifferentialEquations.jl. To use this functionality, you must install DiffEqSensitivity.jl:

]add DiffEqSensitivity using DiffEqSensitivity

You might want to update your docs. Not sure where your snippet is located.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Vaibhavdixit02** [comment]
https://docs.juliadiffeq.org/latest/analysis/global_sensitivity/#GSA-examples-1 leads me to 

![Screenshot 2020-03-20 at 11 10 18 AM](https://user-images.githubusercontent.com/23134958/77139319-8ad65980-6a9b-11ea-9b80-75b3c5ac0a7d.png)

which seems to me is what I was saying. For slack  slackinvite.julialang.org, it is mentioned in Community section in the Julialang website (https://julialang.org/community/)

**Comment by vijaya65** [comment]
Hello
In an earlier example:

m = DiffEqSensitivity.morris_sensitivity(prob,Tsit5(),t,[[1,5],[0.5,5]],[10,10],len_trajectory=1500,total_num_trajectory=1000,num_trajectory=150)

This gives the means of the effects and it's variances over the entire timespan and thus we get 200-length arrays for each paramter and dependent variable pair.

We can plot the trajectory of the sensitivity with the standard deviation bars.


The above command does not work for me.


 With the new construct

m = gsa(f1,Morris(total_num_trajectory=1000,num_trajectory=150),[[1,5],[1,5],[1,5],[1,5]])

How do I define f1 so I get

  200-length arrays for each parameter and dependent variable pair.


Thanks

Vijaya


________________________________

**Comment by Vaibhavdixit02** [comment]
You can change the function to return the solution,

```
f1 = function (p) prob1 = remake(prob;p=p) sol =
solve(prob1,Tsit5();saveat=t) return sol
end
```

**Comment by vijaya65** [comment]
I get the following error when I do that followed by:   m = gsa(f1,Morris(total_num_trajectory=1000,num_trajectory=150),[[1,5],[1,5],[1,5],[1,5]])



DimensionMismatch("tried to assign 2×200 array to 2×200 destination")
in top-level scope at test2.jl:261
in gsa at DiffEqSensitivity\qOHEV\src\global_sensitivity\morris_sensitivity.jl:69
in #gsa#115 at DiffEqSensitivity\qOHEV\src\global_sensitivity\morris_sensitivity.jl:87
in reduce at base\abstractarray.jl:1378
in _typed_hcat at base\abstractarray.jl:1343
in setindex! at base\abstractarray.jl:1074
in _setindex! at base\multidimensional.jl:712
in _unsafe_setindex! at base\multidimensional.jl:717
in macro expansion at base\multidimensional.jl:722
in setindex_shape_check at base\indices.jl:226
in throw_setindex_mismatch at base\indices.jl:169


Thanks
Vijaya

________________________________

**Comment by Vaibhavdixit02** [comment]
Can you provide the entire script that you are trying to run?

**Comment by vijaya65** [comment]
function f(du,u,p,t)
  du[1] = p[1]*u[1] - p[2]*u[1]*u[2] #prey
  du[2] = -p[3]*u[2] + p[4]*u[1]*u[2] #predator
end

u0 = [1.0;1.0]
tspan = (0.0,10.0)
p = [1.5,1.0,3.0,1.0]
prob = ODEProblem(f,u0,tspan,p)
t = collect(range(0, stop=10, length=200))

f1 = function (p)
  prob1 = remake(prob;p=p)
  sol = solve(prob1,Tsit5();saveat=t)
  return sol
#[mean(sol[1,:]),maximum(sol[2,:])]
end

m = gsa(f1,Morris(total_num_trajectory=1000,num_trajectory=150),[[1,5],[1,5],[1,5],[1,5]])



________________________________

**Comment by Vaibhavdixit02** [comment]
One hacky way around it is to do 

```
f1 = function (p)
  prob1 = remake(prob;p=p)
  sol = solve(prob1,Tsit5();saveat=t)
  return vec(sol)
end

m = gsa(f1,Morris(total_num_trajectory=1000,num_trajectory=150),[[1,5],[1,5],[1,5],[1,5]])

m.means

reshape(m.means[:,1],(2,200)) #the  the trajectory of the sensitivity with respect to first parameter

reshape(m.means[:,2],(2,200)) #the  the trajectory of the sensitivity with respect to second parameter and so on
```

I am looking into why it fails otherwise, thanks for reporting the issue.
